Add designer attribute tests

weltkante opened the pull request:
Add tests to ensure that strings specified in designer attributes correspond to existing metadata

Contributes to #2185, #2172, #1115 

## Proposed changes

- Attributes may specify strings to refer to types and other metadata (properties, events). Make sure these actually still exist in the product. These tests help preventing dead references when porting from Desktop to Core and also can detect regressions in future refactoring when renaming something which is technically internal but forgetting that it may have been referenced by attribute strings.

## Customer Impact

- Raise product quality by detecting dead metadata references in attributes

## Regression? 

- Yes (most of the references which are dead now did exist in Desktop Framework)

## Risk

- None, the change is test-only and exists only to detect regressions.

### Before

- Some types and properties carry attributes which refer to nonexisting metadata. These go by undetected until reported by someone.

### After

- Attributes referring to nonexistant metadata should be detected during testing.


## Test methodology <!-- How did you ensure quality? -->

- Manual lookup of attributes which may refer to other metadata by name. May have missed some but the list can be extended anytime.


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2230)

weltkante commented:
/CC: @RussKie @hughbe I'm pretty new to writing tests, critique and suggestions to improve are welcome. Also I don't know how to proceed since there obviously are still a whole bunch of dead attributes which now are flagged as test failures. Some may be known some may be new to the list. There's one dead TypeConverterAttribute (ControlBindingsConverter) and I don't know if anyone actually looked at DesignerAttribute yet, I don't remember any github issue about those.

hughbe commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Design/DesignerAttributeTests.cs:
```diff
@@ -0,0 +1,105 @@
+        [MemberData(nameof(GetAttributeWithType_TestData), AssemblyRef_SystemWinforms, typeof(DefaultBindingPropertyAttribute))]
+        public void DesignerAttributes_DefaultBindingPropertyAttribute_PropertyExists(Type type, DefaultBindingPropertyAttribute attribute)
+        {
+            Assert.NotNull(type.GetProperty(attribute.Name));
```
Can we make these more specific by validating return result?

weltkante replied:
Do you think `Assert.Equals(type.GetProperty(attribute.Name)?.Name, attribute.Name)` would make sense? Then you'd have the attribute name that failed in the test results. I'm not entirely sure which of both should be in the expected and which in the actual parameter though.

If you mean some other kind of validation, thats probably hard to do because in the failure cause you just get `null`. I'm not trying to validate reflection here, I'm just validating if the referenced metadata _exists_.

RussKie commented:
Thank you!
Could you please rebase the PR on release/3.1 branch, because this is where we are currently doing all the work and these test will be quite invaluable there.

We also need to do a bit more work - until we completed the porting, we need to add a list of exclusions, that should start failing, once missing types are ported across.


The current level logging is inadequate (xUnit's fault), so when tests fail it is impossible to say what went wrong. All it tells what line it failed at.
![image](https://user-images.githubusercontent.com/4403806/67739476-b0b87e80-fa66-11e9-9a2c-486cf058cbce.png)

I pushed few tweaks to your PR that increase a level of logging and provide insights into what caused failures (still far from useable, but at least something):
![image](https://user-images.githubusercontent.com/4403806/67739669-59ff7480-fa67-11e9-95e5-86fb9df429ec.png)

weltkante commented:
I hope I rebased correctly, will look into skipping known failures later and sync with the existing issues (or create new issues if they don't exist).

Do you know anything about the `DesignerAttribute` failures? see question [above](https://github.com/dotnet/winforms/pull/2230#issuecomment-547118463):
> I don't know if anyone actually looked at DesignerAttribute yet, I don't remember any github issue about those - or maybe I'm testing them wrong because the assembly isn't part of the WinForms build so lookup can't succeed?

Also there's a test failure here:
> Could not load file or assembly 'System.Windows.Forms.Design, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

I was referencing AssemblyRef.WinformsDesign because there's a single DesignerSerializerAttribute in it. If this test assembly doesn't have access to WinformsDesign then this probably needs to be removed and replaced by an explicit check in some other test assembly.

weltkante commented:
I've been looking into skipping Theory/MemberData items but didn't come up with anything, did I miss something? Do you have an established process to declare which items to skip?

JuditRose commented:
> Also there's a test failure here:
> 
> Could not load file or assembly 'System.Windows.Forms.Design, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

AFAIK ```System.Windows.Forms.Design.Editors``` doesn't depend on ```System.Windows.Forms.Design```, but the other way around. So I'd move ```System.Windows.Forms.Design``` related tests to ```System.Windows.Forms.Design.Tests```. I think this is the same reason ```DesignerAttribute``` cannot be found, so if you move the designer related tests into the designer test assembly, it should work.

weltkante commented:
> AFAIK System.Windows.Forms.Design.Editors doesn't depend on System.Windows.Forms.Design, but the other way around. So I'd move System.Windows.Forms.Design related tests to System.Windows.Forms.Design.Tests. I think this is the same reason DesignerAttribute cannot be found, so if you move the designer related tests into the designer test assembly, it should work.

I suspected as much and did try moving the test, but it didn't change anything (the assembly still wasn't loading), so I didn't push it. I can take another look in a few days.

JuditRose commented:
> I've been looking into skipping Theory/MemberData items but didn't come up with anything, did I miss something? Do you have an established process to declare which items to skip?

I don't think we have any best practices here yet. I'd declare a list of non-ported editors and skip them in the test.

RussKie commented:
We need to rebase on the master branch, 3.1 is complete and closed. Apologies for the hassle.

RussKie commented:
I rebased on the master and f-pushed into your branch.

One thing immediately comes out of these failed tests - the list of types we haven't type-forwarded (TF). We need to TF all internal types that previously lived in other assemblies (e.g. System.Design).
We need to add all the missing TFs, and once we've done that we can look at what types we are missing and can discuss how to exclude them.

@weltkante  @JuditRose do you think you could collaborate and bridge the TF gaps first?

JuditRose commented:
Sure! @weltkante You can see examples how type forwarding is used in Winforms here: https://github.com/dotnet/winforms/blob/master/src/System.Design/src/System.Design.Forwards.cs
Feel free to let me know if you have questions!

weltkante commented:
I can no longer load the WinForms project in VS, I have updated VS 2019 preview and .NET 5 nightly to the latest public versions, as well as executed the build-local.ps1 script, but VS complains about missing SDK 5.0.100-alpha1-013788 (which is older than what I've installed, 5.0.100-alpha1-015756). I made sure I'm working on a clean git clone. I assume I have to wait for the next public preview/release drop of VS for this to work again?

JuditRose commented:
If built from command line sometimes there are lines like these:
```
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-alpha.1.13788.....zip
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-alpha.1.13788.....zip
dotnet-install: Installation finished
```
In this case the dotnet version Winforms uses is not installed locally (and VS doesn't see it, so it can cause errors during build in VS). Usually I copy the link, replace ```zip``` at the end with ```exe``` and paste it in a browser, download it, install it, restart, and VS should be fine afterwards.
I hope this helps!

weltkante commented:
Thanks for the hint, its a bit confusing though, the build downloads `5.0.100-alpha1-015536` but wants to compile against `5.0.100-alpha1-013788`. I had to uninstall the nightly `5.0.100-alpha1-015756` in order to install 15536 matching the build script download, but at that point the build scripts themselves no longer work because they want 13788. Ignoring what the build script downloads and installing 13788 seems to work though. Very weird.

Anyways, I'm back to a working setup now and looking into the type forwardings (seems straightforward) and adding some skipping logic

weltkante commented:
Tests passing, not too happy with the skipping logic but couldn't come up with anything better in short term. Need to remember to remove skipped entries as things are fixed.

I've created new issues for things that weren't covered yet.

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Design/DesignerAttributeTests.cs:
```diff
@@ -0,0 +1,222 @@
+            "System.Windows.Forms.Design.TreeViewDesigner, System.Design, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
+            "System.Windows.Forms.Design.UpDownBaseDesigner, System.Design, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
+            "System.Windows.Forms.Design.UserControlDocumentDesigner, System.Design, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
+            "System.Windows.Forms.Design.WebBrowserDesigner, System.Design, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
```
@merriemcgaw  in what timeframe do we have an intention to port these API across?

RussKie commented:
Awesome!

> thanks for the hint, its a bit confusing though, the build downloads `5.0.100-alpha1-015536` but wants to compile against `5.0.100-alpha1-013788`. I had to uninstall the nightly `5.0.100-alpha1-015756` in order to install 15536 matching the build script download

Nightly builds are incoherent, so things often do not work expected (like you can't build 3.x apps with 5.0 installed).
Sometimes I have work on a clean VM just to make sure things that claim broken are broken indeed. 

> So I'd move `System.Windows.Forms.Design` related tests to `System.Windows.Forms.Design.Tests`. I think this is the same reason `DesignerAttribute` cannot be found, so if you move the designer related tests into the designer test assembly, it should work.

I agree, we are testing types that are declared in System.Windows.Forms.Design.dll, so it feels appropriate to place the tests in the matching test space.
I just realised we have "Design" namespace in System.Windows.Forms.dll, and it is 🤯
![image](https://user-images.githubusercontent.com/4403806/69393859-7329d700-0d2e-11ea-9c8e-5f8641f6a214.png)

weltkante commented:
> I agree, we are testing types that are declared in System.Windows.Forms.Design.dll, so it feels appropriate to place the tests in the matching test space.

See my [last comment](https://github.com/dotnet/winforms/pull/2230#issuecomment-557320159) why I think moving it into the test assembly for System.Windows.Forms.Design is not appropriate. We are not testing the types, we are testing the attributes, and those are not in the Design DLL.

As said, I can still move if you think its better placed there, but I think its appropriate where it is now.

JuditRose commented:
> See my [last comment](https://github.com/dotnet/winforms/pull/2230#issuecomment-557320159) why I think moving it into the test assembly for System.Windows.Forms.Design is not appropriate. We are not testing the types, we are testing the attributes, and those are not in the Design DLL.
> As said, I can still move if you think its better placed there, but IMHO its also appropriate where it is now.

If they are now working with type forwarding then I don't mind keeping them as they are so all related tests are in one place.

RussKie commented:
@hughbe do you have further objections?

hughbe commented:
None!

RussKie commented:
Thank you!
